src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs

```diff
@@ -192,39 +198,38 @@ private void CommonInit()
                 }
                 else
                 {
-                    SearchResult searchResult = CreateSearchResult(_searchData, data);
-                    if (_resultHandler.IsResultIncluded(searchResult))
+                    TSource result;
+                    if (IsResultIncluded(ref data, out result))
                     {
-                        current = _resultHandler.CreateObject(searchResult);
+                        current = result;
                     }
                 }
             }
-            // for AllDirectories, we first recurse into dirs, so cleanup and add searchData 
+            // for AllDirectories, we first recurse into dirs, so cleanup and add searchData
             // to the stack
             else
             {
                 _hnd.Dispose();
+                _searchStack = new List<SearchData>();
```

Comment:
Interesting.  It looks like Array.Reverse tries to special case some types of T, but will fail for all but a known set of primitive types (https://github.com/dotnet/coreclr/blob/4cf8a6b082d9bb1789facd996d8265d3908757b2/src/classlibnative/bcltype/arrayhelpers.cpp#L422-L424).  It'd be nice if in the future we could expose some more generic methods on Array (e.g. ```Reverse<T>```), but in the interim it seems like we should just rewrite ```List<T>.Reverse``` to not use ```Array.Reverse``` (either by just reimplementing ```List<T>.Reverse``` or by adding an internal ```Array.Reverse<T>``` that ```List<T>.Reverse``` could use... in the age of generics, there doesn't seem to be anything particularly magical about the ```TrySZReverse``` implementation in the runtime, anyway. @jkotas, any reason not to do this?